cmd/kubeadm/app/phases/selfhosting/selfhosting.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package selfhosting
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kuberuntime "k8s.io/apimachinery/pkg/runtime"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+)
+
+const (
+	kubeAPIServer         = "kube-apiserver"
+	kubeControllerManager = "kube-controller-manager"
+	kubeScheduler         = "kube-scheduler"
+)
+
+func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+	components := []string{kubeAPIServer, kubeControllerManager, kubeScheduler}
+
+	for _, component := range components {
+		start := time.Now()
+
+		ds, err := buildDaemonSet(component)
+		if err != nil {
+			return err
+		}
+
+		if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+			return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", component, err)
+		}
+
+		// Wait for the self-hosted component to come up
+		kubeadmutil.WaitForPodsWithLabel(client, component, fmt.Sprintf("k8s-app=self-hosted-%s", component))
+
+		manifestPath := buildStaticManifestFilepath(component)
```

Comment:
Could you comment this, it's totally unclear to anyone not working in this area what you are doing, and reference your phases.